crates/zkvm-jetpack/src/jets/mega_jets.rs

```diff
@@ -0,0 +1,165 @@
+    let sam = slot(subject, 6)?;
+    let stack = &mut context.stack;
+
+    let [p_noun, trace_evals_noun, height_noun, chal_map_noun, dyns_noun, com_map_noun] =
+        sam.uncell()?;
+
+    let Ok(trace_evals) = BPolySlice::try_from(trace_evals_noun) else {
+        return jet_err::<Noun>();
+    };
+    let Ok(height_atom) = height_noun.as_atom() else {
+        return jet_err::<Noun>();
+    };
+    let Ok(height) = height_atom.as_u64() else {
+        return jet_err::<Noun>();
+    };
+    let height_usize = height as usize;
+
+    let Ok(dyns) = BPolySlice::try_from(dyns_noun) else {
+        return jet_err::<Noun>();
+    };
+
+    let Ok(chal_map) = HoonMap::try_from(chal_map_noun) else {
+        return jet_err::<Noun>();
+    };
+    let Ok(com_map) = HoonMap::try_from(com_map_noun) else {
+        return jet_err::<Noun>();
+    };
+
+    let mut acc_vec = zero_bpoly();
+
+    let mut p_iter = HoonMapIter::from(p_noun);
+    p_iter.try_fold((), |_, n| {
+        let [k_noun, v_noun] = n.uncell()?;
+        let Ok(k) = BPolySlice::try_from(k_noun) else {
+            return jet_err::<()>();
+        };
+        let Ok(v) = v_noun.as_belt() else {
+            return jet_err::<()>();
+        };
+
+        if v.0 == 0 {
+            return Ok(());
+        }
+
+        let poly_len_for_var_com = 4 * height_usize;
+
+        let mut inner_acc_vec = ones_bpoly(k.len());
+
+        for i in 0..k.len() {
+            let ter = k.0[i];
+
+            let (typ, idx, exp) = brek(ter);
+
+            match typ {
+                MegaTyp::Var => {
+                    let var_start_idx = idx * poly_len_for_var_com;
+                    let var_end_idx = var_start_idx + poly_len_for_var_com;
+
+                    if var_end_idx > trace_evals.len() {
+                        return jet_err::<()>();
+                    }
+                    let var_slice = &trace_evals.0[var_start_idx..var_end_idx];
+
+                    let hadamard_res_len = inner_acc_vec.len().min(var_slice.len());
+                    let (_res_atom, res_poly_slice): (IndirectAtom, &mut [Belt]) =
+                        new_handle_mut_slice(stack, Some(hadamard_res_len));
+
+                    for _ in 0..exp {
+                        let current_inner_acc_slice = &inner_acc_vec.0;
+                        bp_hadamard(current_inner_acc_slice, var_slice, res_poly_slice);
+                        inner_acc_vec = BPolyVec::from(res_poly_slice.iter().map(|&b| b.0).collect::<Vec<u64>>());
+                    }
+                }
+                MegaTyp::Rnd => {
+                    let rnd_noun = chal_map.get(stack, D(idx as u64)).ok_or_else(|| jet_err::<()>().unwrap_err())?;
+                    let Ok(rnd) = rnd_noun.as_belt() else {
+                        return jet_err::<()>();
+                    };
+
+                    let pow_rnd = bpow(rnd.0, exp);
+                    let (_res_atom, res_poly_slice): (IndirectAtom, &mut [Belt]) =
```

Comment:
Same here and with other bpoly's. Need to use this func only when you want to return the value to nockvm